\documentclass[10pt]{article}
\usepackage[margin=1in]{geometry}
\usepackage{longtable}
\usepackage[pdfborder={0 0 0},hyperfootnotes=false]{hyperref}

\newcommand{\mailtourl}[1]{\href{mailto:#1}{\tt #1}}
\newcommand{\tagvalue}[1]{\tt #1}
\newcommand{\tagregex}[1]{\tt #1}

\begin{document}

\input{SAMtags.ver}
\title{Sequence Alignment/Map Optional Fields Specification}
\author{The SAM/BAM Format Specification Working Group}
\date{\headdate}
\maketitle
\begin{quote}\small
The master version of this document can be found at
\url{https://github.com/samtools/hts-specs}.\\
This printing is version~\commitdesc\ from that repository,
last modified on the date shown above.
\end{quote}
\vspace*{1em}

\noindent
This document is a companion to the {\sl Sequence Alignment/Map Format
Specification} that defines the SAM and~BAM formats, and to the {\sl CRAM
Format Specification} that defines the CRAM format.\footnote{See
\href{http://samtools.github.io/hts-specs/SAMv1.pdf}{\tt SAMv1.pdf} and
\href{http://samtools.github.io/hts-specs/CRAMv3.pdf}{\tt CRAMv3.pdf}
at \url{https://github.com/samtools/hts-specs}.}
Alignment records in each of these formats may contain a number of optional
fields, each labelled with a {\it tag\/} identifying that field's data.
This document describes each of the predefined standard tags, and discusses
conventions around creating new tags.

\section{Standard tags}

Predefined standard tags are listed in the following table and described
in greater detail in later subsections.
Optional fields are usually displayed as {\tt TAG:TYPE:VALUE}; the {\it type\/}
may be one of
{\tt A} (character),
{\tt B} (general array),
{\tt f} (real number),
{\tt H} (hexadecimal array),
{\tt i} (integer),
or
{\tt Z} (string).

\begin{center}\small
\begin{longtable}{ccp{12.5cm}}
  \hline
  {\bf Tag} & {\bf Type} & {\bf Description} \\
  \hline
  {\tt AM} & i & The smallest template-independent mapping quality of segments in the rest \\
  {\tt AS} & i & Alignment score generated by aligner \\
  {\tt BA} & B,S & Barcode Calls (per-ZMW) \\
  {\tt BC} & Z & Barcode sequence \\
  {\tt BR} & i & Barcode Quality (per-ZMW) \\
  {\tt BQ} & Z & Offset to base alignment quality (BAQ) \\
  {\tt CC} & Z & Reference name of the next hit \\
  {\tt CM} & i & Edit distance between the color sequence and the color reference (see also {\tt NM})\\
  {\tt CO} & Z & Free-text comments \\
  {\tt CP} & i & Leftmost coordinate of the next hit \\
  {\tt CQ} & Z & Color read base qualities \\
  {\tt CS} & Z & Color read sequence \\
  {\tt CT} & Z & Complete read annotation tag, used for consensus annotation dummy features.\\
  {\tt CX} & i & Subread local context flags.\\
  {\tt DQ} & Z & DeletionQV\\
  {\tt DT} & Z & DeletionTag\\
  {\tt E2} & Z & The 2nd most likely base calls \\
  {\tt FI} & i & The index of segment in the template \\
  {\tt FS} & Z & Segment suffix \\
  {\tt FZ} & B,S & Flow signal intensities \\
  {\tt GC} & ? & Reserved for backwards compatibility reasons \\
  {\tt GQ} & ? & Reserved for backwards compatibility reasons \\
  {\tt GS} & ? & Reserved for backwards compatibility reasons \\
  {\tt H0} & i & Number of perfect hits \\
  {\tt H1} & i & Number of 1-difference hits (see also {\tt NM}) \\
  {\tt H2} & i & Number of 2-difference hits \\
  {\tt HI} & i & Query hit index \\
  {\tt HN} & i & ZMW hole number. \\
  {\tt IH} & i & Number of stored alignments in SAM that contains the query in the current record\\
  {\tt IP} & B,C or B,S & IPD (raw frames or codec V1)\\
  {\tt IQ} & Z & Insertion QV\\  {\tt LB} & Z & Library \\
  {\tt MC} & Z & CIGAR string for mate/next segment\\
  {\tt MD} & Z & String for mismatching positions \\
  {\tt MF} & ? & Reserved for backwards compatibility reasons \\
  {\tt MQ} & i & Mapping quality of the mate/next segment \\
  {\tt MR} & Z & MergeQV \\
  {\tt NH} & i & Number of reported alignments that contains the query in the current record\\
  {\tt NM} & i & Edit distance to the reference \\
  {\tt NP} & i & Number of Passes (1 for subreads, variable for CCS—encodes number of complete passes of the insert).\\
  {\tt OC} & Z & Original CIGAR \\
  {\tt OP} & i & Original mapping position \\
  {\tt OQ} & Z & Original base quality \\
  {\tt PG} & Z & Program \\
  {\tt PQ} & i & Phred likelihood of the template \\
  {\tt PT} & Z & Read annotations for parts of the padded read sequence \\
  {\tt PU} & Z & Platform unit \\
  {\tt PW} & B,C or B,S & PulseWidth (raw frames or codec V1; PacBio-internal) \\
  {\tt QE} & i & 0-based end of query in the ZMW read (absent in CCS)\\
  {\tt QS} & i & 0-based start of query in the ZMW read (absent in CCS)\\
  {\tt QT} & Z & Barcode ({\tt BC} or {\tt RT}) phred-scaled base qualities \\
  {\tt Q2} & Z & Phred quality of the mate/next segment sequence in the {\tt R2} tag \\
  {\tt R2} & Z & Sequence of the mate/next segment in the template \\
  {\tt RG} & Z & Read group \\
  {\tt RQ} & f & Expected accuracy.\\
  {\tt RT} & Z & Barcode sequence (deprecated; use {\tt BC} instead) \\
  {\tt SA} & Z & Other canonical alignments in a chimeric alignment \\
  {\tt SC} & A & Scrap region-type annotation, one of A:=Adapter, B:=Barcode, L:=LQRegion, or F:=Filtered. SC tags ‘A’, ‘B’, and ‘L’ denote specific classes of non-subread data, whereas the ‘F’ tag is reserved for subreads that are undesirable for downstream analysis, e.g., being artifactual or too short. \\
  {\tt SM} & i & Template-independent mapping quality \\
  {\tt SN} & B,f & 4 floats for the average signal-to-noise ratio of A, C, G, and T (in that order) over the HQRegion \\
  {\tt SQ} & ? & Reserved for backwards compatibility reasons \\
  {\tt SR} & Z & SubstitutionQV\\
  {\tt ST} & Z & SubstitutionTag\\
  {\tt S2} & ? & Reserved for backwards compatibility reasons \\
  {\tt SZ} & A & ZMW classification annotation, one of N:=Normal, C:=Control, M:=Malformed, or S:=Sentinel. Reads in the subreads/hqregions/zmws.bam file are implicitly marked as Normal, as they stem from user-defined templates.\\
  {\tt TC} & i & The number of segments in the template \\
  {\tt U2} & Z & Phred probility of the 2nd call being wrong conditional on the best being wrong \\
  {\tt UQ} & i & Phred likelihood of the segment, conditional on the mapping being correct \\
  {\tt X?} & ? & Reserved for end users \\
  {\tt Y?} & ? & Reserved for end users \\
  {\tt Z?} & ? & Reserved for end users \\
  \hline
\end{longtable}
\end{center}

\stepcounter{footnote}
\footnotetext{The IP and PW tags can be encoded in one of two ways either: losslessly as an array of uint16; or in a lossy uint8 array known to PacBio as "codec V1".}

\subsection{Additional Template and Mapping data}

\begin{description}
\item[AM:i:\tagvalue{int}]
The smallest template-independent mapping quality of segments in the rest.

\item[AS:i:\tagvalue{score}]
Alignment score generated by aligner.

\item[BQ:Z:\tagvalue{qualities}]
Offset to base alignment quality (BAQ), of the same length as the read sequence.
At the $i$-th read base, ${\rm BAQ}_i=Q_i-({\rm BQ}_i-64)$ where $Q_i$ is the $i$-th base quality.

\item[CC:Z:\tagvalue{rname}]
Reference name of the next hit; `{\tt =}' for the same chromosome.

\item[CP:i:\tagvalue{pos}]
Leftmost coordinate of the next hit.

\item[E2:Z:\tagvalue{qualities}]
The 2nd most likely base calls. Same encoding and same length as {\sf QUAL}.

\item[FI:i:\tagvalue{int}]
The index of segment in the template.

\item[FS:Z:\tagvalue{str}]
Segment suffix.

\item[H0:i:\tagvalue{count}]
Number of perfect hits.

\item[H1:i:\tagvalue{count}]
Number of 1-difference hits (see also {\tt NM}).

\item[H2:i:\tagvalue{count}]
Number of 2-difference hits.

\item[HI:i:\emph{i}]
Query hit index, indicating the alignment record is the $i$-th one stored
in SAM.

\item[IH:i:\tagvalue{count}]
Number of stored alignments in SAM that contains the query in the current
record.

\item[MC:Z:\tagvalue{cigar}]
CIGAR string for mate/next segment.

\item[MD:Z:\tagregex{[0-9]+(([A-Z]|\char92\char94[A-Z]+)[0-9]+)*}]
String for mismatching positions.

The {\tt MD} field aims to achieve SNP/indel calling without
looking at the reference. For example, a string `{\tt 10A5\char94AC6}' means
from the leftmost reference base in the alignment, there are 10 matches
followed by an A on the reference which is different from the aligned read
base; the next 5 reference bases are matches followed by a 2bp deletion from
the reference; the deleted sequence is AC; the last 6~bases are matches.
The {\tt MD} field ought to match the {\sf CIGAR} string.

\item[MQ:i:\tagvalue{}]
Mapping quality of the mate/next segment.

\item[NH:i:\tagvalue{}]
Number of reported alignments that contains the query in the current record.

\item[NM:i:\tagvalue{}]
Edit distance to the reference, including ambiguous bases but excluding clipping.

\item[PQ:i:\tagvalue{}]
Phred likelihood of the template, conditional on both the mapping being correct.

\item[Q2:Z:\tagvalue{}]
Phred quality of the mate/next segment sequence in the {\tt R2} tag.
Same encoding as {\sf QUAL}.

\item[R2:Z:\tagvalue{}]
Sequence of the mate/next segment in the template.

\item[SA:Z:\tagregex{{\tt (}\emph{rname}{\tt ,}\emph{pos}{\tt ,}\emph{strand}{\tt ,}\emph{CIGAR}{\tt ,}\emph{mapQ}{\tt ,}\emph{NM}{\tt ;)}+}]
Other canonical alignments in a chimeric alignment, formatted as a semicolon-delimited list.
Each element in the list represents a part of the chimeric alignment. Conventionally, at a supplementary line, the first element points to the primary line.

\item[SM:i:\tagvalue{}]
Template-independent mapping quality.

\item[TC:i:\tagvalue{}]
The number of segments in the template.

\item[U2:Z:\tagvalue{}]
Phred probility of the 2nd call being wrong conditional on the best being wrong.
The same encoding as {\sf QUAL}.

\item[UQ:i:\tagvalue{}]
Phred likelihood of the segment, conditional on the mapping being correct.
\end{description}

\subsection{Metadata}

\begin{description}
\item[RG:Z:\tagvalue{readgroup}]
The read group to which the read belongs.
If {\tt @RG} headers are present, then \emph{readgroup} must match the
{\tt RG-ID} field of one of the headers.

\item[LB:Z:\tagvalue{library}]
The library from which the read has been sequenced.
If {\tt @RG} headers are present, then \emph{library} must match the
{\tt RG-LB} field of one of the headers.

\item[PG:Z:\tagvalue{}]
Program. Value matches the header {\tt PG-ID} tag if {\tt @PG} is present.

\item[PU:Z:\tagvalue{platformunit}]
The platform unit in which the read was sequenced.
If {\tt @RG} headers are present, then \emph{platformunit} must match the
{\tt RG-PU} field of one of the headers.

\item[CO:Z:\tagvalue{text}]
Free-text comments.
\end{description}

\subsection{Barcodes}

\begin{description}
\item[BC:Z:\tagvalue{sequence}]
Barcode sequence, with any quality scores stored in the {\tt QT} tag.

\item[QT:Z:\tagvalue{qualities}]
Phred quality of the barcode sequence in the {\tt BC} (or {\tt RT}) tag. Same encoding as {\sf QUAL}.

\item[RT:Z:\tagvalue{sequence}]
Deprecated alternative to {\tt BC} tag originally used at Sanger.
\end{description}

\subsection{Original data}

\begin{description}
\item[OC:Z:\tagvalue{cigar}]
Original CIGAR, usually before realignment.

\item[OP:i:\tagvalue{pos}]
Original mapping position, usually before realignment.

\item[OQ:Z:\tagvalue{qualities}]
Original base quality, usually before recalibration.
Same encoding as {\sf QUAL}.
\end{description}

\subsection{Annotation and Padding}

\begin{description}
\item[CT:Z:\tagregex{\emph{strand};\emph{type}(;\emph{key}(=\emph{value}))*}]
Complete read annotation tag, used for consensus annotation dummy features.

The {\tt CT} tag is intended primarily for annotation
dummy reads, and consists of a \emph{strand}, \emph{type} and zero or
more \emph{key}=\emph{value} pairs, each separated with semicolons.
The \emph{strand} field has four values as in GFF3, and supplements FLAG
bit 0x10 to allow unstranded (`{\tt .}'), and stranded but unknown strand
(`{\tt ?}') annotation. For these and annotation on the forward strand
(\emph{strand} set to `{\tt +}'), do not set FLAG bit 0x10. For
annotation on the reverse strand, set the \emph{strand} to `{\tt -}'
and set FLAG bit 0x10.

The \emph{type} and any \emph{keys} and their
optional \emph{values} are all percent encoded according to
RFC3986 to escape meta-characters `{\tt =}', `{\tt \%}', `{\tt ;}',
`{\tt |}' or non-printable characters not matched by the isprint()
macro (with the C locale). For example a percent sign becomes
`{\tt \%2C}'.
%NOTE - This leaves open the possibility of allowing multiple such
%entries for a single CT tag to be combined with | as in the PT tag.

\item[PT:Z:\tagregex{\tt \emph{start};\emph{end};\emph{strand};\emph{type}(;\emph{key}(=\emph{value}))*(\char92|\emph{start};\emph{end};\emph{strand};\emph{type}(;\emph{key}(=\emph{value}))*)*}]
Read annotations for parts of the padded read sequence.

The {\tt PT} tag value has the format of a series of
tags separated by `{\tt |}', each annotating a sub-region of the read.
Each tag consists of \emph{start}, \emph{end}, \emph{strand},
\emph{type} and zero or more \emph{key}{\tt =}\emph{value} pairs, each
separated with semicolons. \emph{Start} and \emph{end} are 1-based
positions between one and the sum of the {\tt M/I/D/P/S/=/X}
{\sf CIGAR} operators, i.e. {\sf SEQ} length plus any pads.  Note
any editing of the CIGAR string may require updating the `{\tt PT}'
tag coordinates, or even invalidate them.
As in GFF3, \emph{strand} is one of `{\tt +}' for forward strand tags,
`{\tt -}' for reverse strand, `{\tt .}' for unstranded or `{\tt ?}'
for stranded but unknown strand.
The \emph{type} and any \emph{keys} and their optional \emph{values}
are all percent encoded as in the {\tt CT} tag.
\end{description}

\subsection{Technology-specific data}

\begin{description}
\item[FZ:B,S:\tagvalue{intensities}]
Flow signal intensities on the original strand of the read, stored as {\tt (uint16\_t) round(value * 100.0)}.
\end{description}

\subsubsection{Color space}

% TODO Describe color space and the encoding here.

\begin{description}
\item[CM:i:\tagvalue{distance}]
Edit distance between the color sequence and the color reference (see also {\tt NM}).

\item[CS:Z:\tagvalue{sequence}]
Color read sequence on the original strand of the read. The primer base must be included.

\item[CQ:Z:\tagvalue{qualities}]
Color read quality on the original strand of the read. Same encoding as {\sf QUAL}; same length as {\tt CS}.
\end{description}

\section{Locally-defined tags}

You can freely add new tags.
Note that tags starting with `{\tt X}', `{\tt Y}', or `{\tt Z}' and tags
containing lowercase letters in either position are reserved for local use
and will not be formally defined in any future version of this specification.

If a new tag may be of general interest, it may be useful to have it added
to this specification.  Additions can be proposed by opening a new issue at
\url{https://github.com/samtools/hts-specs/issues} and/or by sending email
to \mailtourl{samtools-devel@lists.sourceforge.net}.

\end{document}
